File: security/src/main/java/net/yan100/compose/security/UserDetailsWrapper.java
package net.yan100.compose.security;

import lombok.extern.slf4j.Slf4j;
import net.yan100.compose.core.lang.Str;
import net.yan100.compose.core.models.AuthUserInfo;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.util.ArrayList;
import java.util.Collection;

/**
 * UserDetailsWrapper 包装类
 *
 * @author TrueNine
 * @since 2022-12-10
 */
@Slf4j
public record UserDetailsWrapper(AuthUserInfo authUserInfo) implements UserDetails {

  public UserDetailsWrapper {
    log.trace("构建 = {}", authUserInfo);
  }

  @Override
  public Collection<? extends GrantedAuthority> getAuthorities() {
    var auths = new ArrayList<GrantedAuthority>();
   // 添加角色信息
    this.authUserInfo.getRoles().forEach(r -> auths.add(new SimpleGrantedAuthority("ROLE_" + r)));
    // 添加权限信息
    this.authUserInfo.getPermissions().forEach(p -> auths.add(new SimpleGrantedAuthority(p)));
    // 添加 部门信息 到鉴权列表
    this.authUserInfo.getDepts().stream().filter(Str::hasText).map(r -> "DEPT_" + r).forEach(r -> auths.add(new SimpleGrantedAuthority(r)));

    return auths;
  }

  @Override
  public String getPassword() {
    return authUserInfo.getEncryptedPassword();
  }

  @Override
  public String getUsername() {
    return authUserInfo.getAccount();
  }

  @Override
  public boolean isAccountNonExpired() {
    return authUserInfo.getNonExpired();
  }

  @Override
  public boolean isAccountNonLocked() {
    return authUserInfo.getNonLocked();
  }

  @Override
  public boolean isCredentialsNonExpired() {
    return authUserInfo.getNonExpired();
  }

  @Override
  public boolean isEnabled() {
    return authUserInfo.getEnabled();
  }
}

File: security/src/main/kotlin/net/yan100/compose/security/UserDetailsWrapper.kt
/*
 *  Copyright (c) 2020-2024 TrueNine. All rights reserved.
 *
 * The following source code is owned, developed and copyrighted by TrueNine
 * (truenine304520@gmail.com) and represents a substantial investment of time, effort,
 * and resources. This software and its components are not to be used, reproduced,
 * distributed, or sublicensed in any form without the express written consent of
 * the copyright owner, except as permitted by law.
 * Any unauthorized use, distribution, or modification of this source code,
 * or any portion thereof, may result in severe civil and criminal penalties,
 * and will be prosecuted to the maximum extent possible under the law.
 * For inquiries regarding usage or redistribution, please contact:
 *     TrueNine
 *     email: <truenine304520@gmail.com>
 *     website: <github.com/TrueNine>
 */
package net.yan100.compose.security

import net.yan100.compose.core.models.AuthRequestInfo
import net.yan100.compose.core.util.Str
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails

/**
 * UserDetailsWrapper 包装类
 *
 * @author TrueNine
 * @since 2022-12-10
 */
data class UserDetailsWrapper(val authUserInfo: AuthRequestInfo?) : UserDetails {
  private val auths: MutableSet<GrantedAuthority> = mutableSetOf()

  init {
    // 添加角色信息
    auths += authUserInfo?.roles?.filter(Str::hasText)?.map { SimpleGrantedAuthority("ROLE_$it") } ?: emptyList()
    // 添加权限信息
    auths += authUserInfo?.permissions?.filter(Str::hasText)?.map { SimpleGrantedAuthority(it) } ?: emptyList()
    // 添加 部门信息
    auths += authUserInfo?.depts?.filter(Str::hasText)?.map { SimpleGrantedAuthority("DEPT_$it") } ?: emptyList()
  }

  override fun getAuthorities(): Collection<GrantedAuthority> = auths

  override fun getPassword(): String = authUserInfo?.encryptedPassword!!

  override fun getUsername(): String = authUserInfo?.account!!

  override fun isAccountNonExpired(): Boolean = authUserInfo?.nonExpired ?: true

  override fun isAccountNonLocked(): Boolean = authUserInfo?.nonLocked ?: true

  override fun isCredentialsNonExpired(): Boolean = authUserInfo?.nonExpired ?: true

  override fun isEnabled(): Boolean = authUserInfo?.enabled ?: true
}
